data_safe_haven/infrastructure/programs/sre/firewall.py

```diff
@@ -111,206 +114,238 @@ def __init__(
+                    ),
+                    network.AzureFirewallApplicationRuleArgs(
+                        description="Allow external RStudio deb downloads",
+                        name="AllowRStudioDeb",
+                        protocols=[
+                            network.AzureFirewallApplicationRuleProtocolArgs(
+                                port=int(Ports.HTTPS),
+                                protocol_type=network.AzureFirewallApplicationRuleProtocolType.HTTPS,
+                            ),
+                        ],
+                        source_addresses=props.subnet_workspaces_prefixes,
+                        target_fqdns=PermittedDomains.RSTUDIO_DEB,
+                    ),
+                ],
+            ),
+            network.AzureFirewallApplicationRuleCollectionArgs(
+                action=network.AzureFirewallRCActionArgs(
+                    type=network.AzureFirewallRCActionType.DENY
                 ),
-                network.AzureFirewallApplicationRuleCollectionArgs(
+                name="workspaces-deny",
+                priority=FirewallPriorities.SRE_WORKSPACES_DENY,
+                rules=[
+                    network.AzureFirewallApplicationRuleArgs(
+                        description="Deny external Ubuntu Snap Store upload and login access",
+                        name="DenyUbuntuSnapcraft",
+                        protocols=[
+                            network.AzureFirewallApplicationRuleProtocolArgs(
+                                port=int(Ports.HTTP),
+                                protocol_type=network.AzureFirewallApplicationRuleProtocolType.HTTP,
+                            ),
+                            network.AzureFirewallApplicationRuleProtocolArgs(
+                                port=int(Ports.HTTPS),
+                                protocol_type=network.AzureFirewallApplicationRuleProtocolType.HTTPS,
+                            ),
+                        ],
+                        source_addresses=props.subnet_workspaces_prefixes,
+                        target_fqdns=ForbiddenDomains.UBUNTU_SNAPCRAFT,
+                    ),
+                ],
+            ),
+        ]
+
+        network_rule_collections: list[
+            network.AzureFirewallNetworkRuleCollectionArgs
+        ] = []
+
+        if props.allow_workspace_internet:
+            application_rule_collections = []
+            network_rule_collections.append(
+                network.AzureFirewallNetworkRuleCollectionArgs(
                     action=network.AzureFirewallRCActionArgs(
-                        type=network.AzureFirewallRCActionType.DENY
+                        type=network.AzureFirewallRCActionType.ALLOW
                     ),
-                    name="workspaces-deny",
-                    priority=FirewallPriorities.SRE_WORKSPACES_DENY,
+                    name="workspaces-all-allow",  # TODO: Fix other names.
```

Comment:
Let's solve this here.

Reply:
I'm not actually sure what this means, I can't see any other instances of this class.

Reply:
I think the suggestion was to add "allow" or "deny" to all firewall rule names.